### PR TITLE
fix for middleware generating empty routes with rails 8

### PR DESCRIPTION
**Pull request by francescob**
this fixes issue #319

### CODE REVIEW

**Comment by le0pard** [comment]
maybe need create gemfile with rails 8 and add it in ci - https://github.com/railsware/js-routes/blob/main/.github/workflows/ci.yml#L19

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -17,6 +17,7 @@ jobs:
           - gemfiles/rails51_sprockets_3.gemfile
           - gemfiles/rails52_sprockets_3.gemfile
           - gemfiles/rails70_sprockets_4.gemfile
+          - gemfiles/rails70_sprockets_4.gemfile
```
Comment (le0pard) [inline .github/workflows/ci.yml]:
```suggestion
          - gemfiles/rails80_sprockets_4.gemfile
```

Comment (francescob) [reply .github/workflows/ci.yml]: yup sorry, that was a typo, already fixed it

**Comment by le0pard** [comment]
```
Because railties >= 8.0.0.beta1 depends on Ruby >= 3.2.0
  and rails80_sprockets_4.gemfile depends on railties ~> 8.0,
  Ruby >= 3.2.0 is required.

```

ok, so maybe lets try this way:

```yml
name: CI

on:
  push:
    branches:
      - master
  pull_request:

jobs:
  build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        include:
          - ruby: "2.7"
            gemfile: gemfiles/rails50_sprockets_3.gemfile
          - ruby: "2.7"
            gemfile: gemfiles/rails51_sprockets_3.gemfile
         - ruby: "2.7"
            gemfile: gemfiles/rails52_sprockets_3.gemfile
         - ruby: "2.7"
            gemfile: gemfiles/rails70_sprockets_4.gemfile
         - ruby: "3.3"
            gemfile: gemfiles/rails80_sprockets_4.gemfile
    env:
      BUNDLE_GEMFILE: ${{ matrix.gemfile }}
    steps:
      - uses: actions/checkout@v4
      - uses: ruby/setup-ruby@v1
        with:
          ruby-version: ${{ matrix.ruby }}
          bundler-cache: true
      - uses: actions/setup-node@v4
        with:
          node-version: "22"
          cache: 'yarn'
      - name: install dependencies
        run: |
          yarn install
      - name: test
        run: bundle exec rake spec
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -11,13 +11,16 @@ jobs:
...
     strategy:
       matrix:
-        ruby: ["2.7"]
-        gemfile:
-          - gemfiles/rails50_sprockets_3.gemfile
-          - gemfiles/rails51_sprockets_3.gemfile
-          - gemfiles/rails52_sprockets_3.gemfile
-          - gemfiles/rails70_sprockets_4.gemfile
-
+        include:
```
Comment (le0pard) [inline .github/workflows/ci.yml:14]: even better

Comment (le0pard) [reply .github/workflows/ci.yml:14]: @francescob try one-by-one like https://github.com/railsware/js-routes/pull/320#issuecomment-2468321447

**Comment by francescob** [comment]
having  a look at that failure now

**Comment by le0pard** [comment]
Looks like no block, but can set like this in rails 8:

```ruby
ActiveSupport::Deprecation.silenced = true
ActiveSupport::Deprecation.behavior = :silence
```

**Comment by le0pard** [comment]
@francescob looks like this way even better:

```ruby
(Rails.try(:deprecator) || ActiveSupport::Deprecation).silence do
  ...
end
```

**Comment by le0pard** [comment]
@francescob great job. Now need to wait for @bogdan review

File: lib/js_routes/instance.rb [hunk 3]
```diff
@@ -24,9 +24,14 @@ def initialize(**options)
...
     def generate
       # Ensure routes are loaded. If they're not, load them.
+
       application = T.unsafe(self.application)
-      if named_routes.empty? && application.respond_to?(:reload_routes!, true)
-        application.reload_routes!
+      if named_routes.empty?
+        if application.respond_to?(:reload_routes_unless_loaded, true)
+          application.reload_routes_unless_loaded
+        elsif application.respond_to?(:reload_routes!, true)
```
Comment (bogdan) [inline lib/js_routes/instance.rb:32]: Is it even possible for `application` to not respond to this method in supported versions of Rails? Not is the scope of this PR, I would be grateful if you can investigate this too.

File: spec/js_routes/options_spec.rb [hunk 4]
```diff
@@ -466,7 +466,7 @@
 
     context "with url_links option" do
       around(:each) do |example|
-        ActiveSupport::Deprecation.silence do
+        (Rails.try(:deprecator) || ActiveSupport::Deprecation).silence do
```
Comment (bogdan) [inline spec/js_routes/options_spec.rb:469]: Same here, prefer `Rails.version` check if possible.

Comment (le0pard) [reply spec/js_routes/options_spec.rb:469]: Idea about such fix was taken from this - https://github.com/rails-api/active_model_serializers/pull/2483/files

File: lib/js_routes/instance.rb [hunk 5]
```diff
@@ -24,9 +24,14 @@ def initialize(**options)
     sig {returns(String)}
     def generate
       # Ensure routes are loaded. If they're not, load them.
+
       application = T.unsafe(self.application)
-      if named_routes.empty? && application.respond_to?(:reload_routes!, true)
-        application.reload_routes!
+      if named_routes.empty?
+        if application.respond_to?(:reload_routes_unless_loaded, true)
```
[the remark this answers is not in the document]
Comment (le0pard) [reply lib/js_routes/instance.rb:30]: `rails8_or_never? = Gem::Version.new(Rails.version) >= Gem::Version.new('8')` looks like needed

File: spec/js_routes/options_spec.rb [hunk 6]
```diff
@@ -466,7 +466,9 @@
 
     context "with url_links option" do
       around(:each) do |example|
-        ActiveSupport::Deprecation.silence do
+        rails8_or_newer = Gem::Version.new(Rails.version) >= Gem::Version.new('8')
+        deprecator_object = rails8_or_newer ? Rails.deprecator : ActiveSupport::Deprecation
```
Comment (bogdan) [inline spec/js_routes/options_spec.rb:470]: `Rails.deprecator` introduced in `7.1`: https://github.com/rails/rails/blob/v7.1.0/railties/lib/rails/deprecator.rb


```suggestion
        Rails.version >= "7.1.0" ? Rails.deprecator : ActiveSupport::Deprecation
```

File: lib/js_routes/instance.rb [hunk 7]
```diff
@@ -24,9 +24,15 @@ def initialize(**options)
     sig {returns(String)}
     def generate
       # Ensure routes are loaded. If they're not, load them.
+
       application = T.unsafe(self.application)
-      if named_routes.empty? && application.respond_to?(:reload_routes!, true)
-        application.reload_routes!
+      if named_routes.empty?
+        rails8_or_newer = Gem::Version.new(Rails.version) >= Gem::Version.new('8')
+        if rails8_or_newer
```
Comment (bogdan) [inline lib/js_routes/instance.rb:31]: String comparison is enough:

```suggestion
        if Rails.version >= "8.0.0"
```

File: lib/js_routes/instance.rb [hunk 8]
```diff
@@ -24,9 +24,15 @@ def initialize(**options)
...
+          application.reload_routes_unless_loaded
+        elsif application.respond_to?(:reload_routes!, true)
+          application.reload_routes!
```
[the remark this answers is not in the document]
Comment (francescob) [reply lib/js_routes/instance.rb:34]: what shall we do in case of a Rails::Engine? just nothing?

**Pull request merged**